### PR TITLE
Arthig/error handling

### PR DESCRIPTION
**Pull request by agokarn**
This PR has the following changes -
1. error handling - removal of panic, error handling in the main thread to get logs.
2. Added file and line number to the logs for debuggability.
3. Did some refactoring for readability and some more documentation

### CODE REVIEW

File: ohttp-client/src/main.rs [hunk 1]
```diff
@@ -176,98 +260,101 @@ struct KmsKeyConfiguration {
...
 
-    let (enc_request, client_response) = ohttp_request.encapsulate(&request_buf)?;
-    info!(
-        "Sending encrypted OHTTP request to {}: {}",
-        args.url,
-        hex::encode(&enc_request[0..60])
-    );
+/// Creates an OHTTP client from KMS.
+///
+async fn create_request_handler_from_kms(
```
Comment (kapilvgit) [inline ohttp-client/src/main.rs:291]: not sure about the use of handler in this name

File: ohttp-client/src/main.rs [hunk 2]
```diff
@@ -176,98 +260,101 @@ struct KmsKeyConfiguration {
...
-        from_kms_config(&config, &cert)?
-    } else {
-        let config = &args.config.clone().expect("Config expected.");
-        ohttp::ClientRequest::from_encoded_config_list(config)?
-    };
-
-    trace!("================== STEP 2 ==================");
+/// Creates an OHTTP client from the static config provided in Args.
+///
+fn create_request_handler(config: &Option<HexArg>) -> Res<ohttp::ClientRequest> {
```
Comment (kapilvgit) [inline ohttp-client/src/main.rs:284]: this is returning a ClientRequest, not a handler.

File: ohttp-server/src/main.rs [hunk 3]
```diff
@@ -160,38 +166,63 @@ async fn import_config(maa: &str, kms: &str) -> Res<KeyConfig> {
...
+    // Get HPKE private key from Azure KMS
+    let client = Client::builder()
+        .danger_accept_invalid_certs(true)
+        .build()?;
+    let key = get_hpke_private_key_from_kms(&client, kms, &token).await?;
+
+    // Parse the returned CBOR key (in CWK-like format)
+    let (kid, d) = parse_cbor_key(&key)?;
+
+    // Get the private and public keys from the encided key
```
Comment (kapilvgit) [inline ohttp-server/src/main.rs:214]: encoded

File: ohttp-server/src/main.rs [hunk 4]
```diff
@@ -207,24 +238,9 @@ async fn import_config(maa: &str, kms: &str) -> Res<KeyConfig> {
...
-    let method: Method = if let Some(method_bytes) = bin_request.control().method() {
-        Method::from_bytes(method_bytes)?
-    } else {
-        Method::GET
-    };
-
-    // Copy headers from the encapsulated request
+/// Copies headers from the encapsulated request and logs them.
+///
+fn copy_headers_from_request(bin_request: &Message) -> HeaderMap {
```
Comment (kapilvgit) [inline ohttp-server/src/main.rs:243]: nit: this is just getting headers, not copying...shall we just call this get_headers_from_request?

File: ohttp-client/src/main.rs [hunk 5]
```diff
@@ -70,49 +74,63 @@ struct Args {
...
+/// Writes the request line for an HTTP POST request to the provided buffer.
+/// The request line follows the format:
+/// `POST {target_path} HTTP/1.1\r\n`.
+fn write_post_request_line(request: &mut Vec<u8>, target_path: &str) -> Res<()> {
+    write!(request, "POST {target_path} HTTP/1.1\r\n")?;
+    Ok(())
+}
 
+/// Appends HTTP headers to the provided request buffer.
+fn append_headers(request: &mut Vec<u8>, headers: &Option<Vec<String>>) -> Res<()> {
```
Comment (kapilvgit) [inline ohttp-client/src/main.rs:103]: nit: some methods are write and some are append

File: ohttp-client/src/main.rs [hunk 6]
```diff
@@ -135,17 +153,84 @@ fn create_multipart_request(
...
+fn create_request_buffer(
+    is_bhttp: bool,
+    target_path: &str,
+    headers: &Option<Vec<String>>,
+    form_fields: &Option<Vec<String>>,
+) -> Res<Vec<u8>> {
+    let request = create_multipart_request(target_path, headers, form_fields)?;
+    let mut cursor = Cursor::new(request);
+
+    // If `is_bhttp` is `true`, it reads a BHTTP message using `Message::read_bhttp`.
```
Comment (kapilvgit) [inline ohttp-client/src/main.rs:221]: Avoid obvious comments.

File: ohttp-client/src/main.rs [hunk 7]
```diff
@@ -176,98 +261,101 @@ struct KmsKeyConfiguration {
 
 /// Reads a json containing key configurations with receipts and constructs
 /// a single use client sender from the first supported configuration.
-pub fn from_kms_config(config: &str, cert: &str) -> Res<ClientRequest> {
+pub fn create_ohttp_request_from_kms_config(config: &str, cert: &str) -> Res<ClientRequest> {
```
Comment (kapilvgit) [inline ohttp-client/src/main.rs:264]: not: ClientRequest is not an OHTTP request. Also the from_kms_config mirrors the from_ functions in the OHTTP library.

File: ohttp-client/src/main.rs (hunk 5 above)
Comment (agokarn) [reply ohttp-client/src/main.rs:103]: @kapilvgit can you please let me know what you want changed?

File: ohttp-client/src/main.rs [hunk 8]
```diff
@@ -176,98 +261,101 @@ struct KmsKeyConfiguration {
...
-        from_kms_config(&config, &cert)?
-    } else {
-        let config = &args.config.clone().expect("Config expected.");
-        ohttp::ClientRequest::from_encoded_config_list(config)?
-    };
-
-    trace!("================== STEP 2 ==================");
+/// Creates an OHTTP client from the static config provided in Args.
+///
+fn create_ohttp_request(config: &Option<HexArg>) -> Res<ohttp::ClientRequest> {
```
Comment (kapilvgit) [inline ohttp-client/src/main.rs:285]: Same as above. Not clear from the function name what is being created.

File: ohttp-client/src/main.rs (hunk 7 above)
Comment (agokarn) [reply ohttp-client/src/main.rs:264]: @kapilvgit  what do you suggest?

File: ohttp-client/src/main.rs (hunk 5 above)
Comment (kapilvgit) [reply ohttp-client/src/main.rs:103]: make this write_headers?

File: ohttp-client/src/main.rs [hunk 9]
```diff
@@ -135,17 +153,82 @@ fn create_multipart_request(
         }
     }
 
+    Ok(body)
+}
+
+/// Append the headers for a multipart/form-data HTTP request to the provided buffer.
+///      Content-Type: multipart/form-data; boundary=---------------------------boundaryString
+///      Content-Length: 12345
+fn append_multipart_headers(request: &mut Vec<u8>, boundary: &str, body_len: usize) -> Res<()> {
```
Comment (kapilvgit) [inline ohttp-client/src/main.rs:162]: And write_multipart_headers

**Review by kapilvgit — APPROVED** [review]

**Pull request merged**